### PR TITLE
PR: Implement Jakob and Hanika (2019) spectral upsampling

### PR DESCRIPTION
**Pull request by enneract**
An implementation of Jakob and Hanika (2019) spectral upsampling (#458). This is a part of my "New Spectral Upsampling Methods" GSoC 2020 project.

### CODE REVIEW

**Comment by KelSolaar** [comment]
Thanks @enneract!

I'm not sure why the checks did not run but I changed the base branch from `feature/jakob2019` to `develop` and will close/open the PR to see if it triggers them!

**Comment by KelSolaar** [comment]
Here we go! :)

File: colour/recovery/tests/test_jakob2019.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+                XYZ = sd_to_XYZ(sd, illuminant=D65) / 100
+
+                RGB = XYZ_to_RGB(
+                    XYZ,
+                    D65_xy,
+                    colourspace.whitepoint,
+                    colourspace.XYZ_to_RGB_matrix,
+                )
+
+                recovered_sd, error = RGB_to_sd_Jakob2019(
```
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:65]: Codacy found an issue: [Unused variable 'recovered_sd'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- colour/recovery/jakob2019.py  11
- colour/recovery/tests/test_jakob2019.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724&bid=18835358)

File: colour/recovery/tests/test_jakob2019.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+        # error_function will not align these for us.
+        shape = DEFAULT_SPECTRAL_SHAPE_JAKOB_2019
+        aligned_cmfs = cmfs.copy().align(shape)
+        illuminant = D65.copy().align(shape)
+        illuminant_XYZ = sd_to_XYZ(D65)
+        illuminant_XYZ /= illuminant_XYZ[1]
+
+        for coefficients in coefficient_list:
+            error, derror, R, XYZ, Lab = error_function(
```
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:115]: Codacy found an issue: [Unused variable 'derror'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- colour/recovery/tests/test_jakob2019.py  4
- colour/recovery/jakob2019.py  11
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724&bid=18835358)

File: colour/recovery/tests/test_jakob2019.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+                XYZ = sd_to_XYZ(sd, illuminant=D65) / 100
+
+                RGB = XYZ_to_RGB(
+                    XYZ,
+                    D65_xy,
+                    colourspace.whitepoint,
+                    colourspace.XYZ_to_RGB_matrix,
+                )
+
+                recovered_sd, error = RGB_to_sd_Jakob2019(
```
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:64]: Codacy found an issue: [Unused variable 'recovered_sd'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

File: colour/recovery/tests/test_jakob2019.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+        # error_function will not align these for us.
+        shape = DEFAULT_SPECTRAL_SHAPE_JAKOB_2019
+        aligned_cmfs = cmfs.copy().align(shape)
+        illuminant = D65.copy().align(shape)
+        illuminant_XYZ = sd_to_XYZ(D65)
+        illuminant_XYZ /= illuminant_XYZ[1]
+
+        for coefficients in coefficient_list:
+            error, derror, R, XYZ, Lab = error_function(
```
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:114]: Codacy found an issue: [Unused variable 'derror'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- colour/recovery/jakob2019.py  11
- colour/recovery/tests/test_jakob2019.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724&bid=18835358)

File: colour/recovery/tests/test_jakob2019.py (hunk 3 above)
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:64]: Codacy found an issue: [Unused variable 'recovered_sd'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

File: colour/recovery/tests/test_jakob2019.py (hunk 4 above)
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:114]: Codacy found an issue: [Unused variable 'derror'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

File: colour/recovery/tests/test_jakob2019.py (hunk 3 above)
Comment (KelSolaar) [reply colour/recovery/tests/test_jakob2019.py:64]: This is actually not me writing that but Codacy.

File: colour/recovery/tests/test_jakob2019.py (hunk 4 above)
Comment (KelSolaar) [reply colour/recovery/tests/test_jakob2019.py:114]: This is actually not me writing that but Codacy.

**Comment by KelSolaar** [comment]
Ah! I was like, where are my two commits, I think I pushed them to your fork instead of this repo branch. Did you give me access rights @enneract? :) Sorry about that!

Anyway, I vectorised the error function, it might be worth document the derivative computation and how they are roughly used.

I did not do the LUT loop yet: https://github.com/colour-science/colour/blob/f2a8baaf28f933cd90e733293812146fcd904273/colour/recovery/jakob2019.py#L516
as it is trickier.

There are quite a few minor things to address but it is taking shape!

**Comment by KelSolaar** [comment]
@enneract : I think that should work for the LUT:

```python
    samples = np.linspace(0, 1, chroma_steps)
    indexes = np.arange(chroma_steps)

    table = np.meshgrid(*[[0, 1, 2], indexes, indexes], indexing='ij')
    chromas = np.transpose(table).reshape(-1, 3)[..., [0, 2, 1]]

    xy = np.meshgrid(*[[1, 1, 1], samples, samples], indexing='ij')
    xy = np.transpose(xy).reshape(-1, 3)[..., [0, 2, 1]]
    indexes = np.tile([[0, 1, 2], [1, 0, 2], [1, 2, 0]], (chroma_steps ** 2, 1))
    RGB = np.take_along_axis(xy, indexes, axis=1)
```

I reckon it would be more elegant if there was no need to reorder the columns of the last axis: `[..., [0, 2, 1]]`

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- colour/recovery/jakob2019.py  11
- colour/recovery/tests/test_jakob2019.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724&bid=18835358)

File: colour/recovery/tests/test_jakob2019.py (hunk 3 above)
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:64]: Codacy found an issue: [Unused variable 'recovered_sd'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

File: colour/recovery/tests/test_jakob2019.py (hunk 4 above)
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:114]: Codacy found an issue: [Unused variable 'derror'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- colour/recovery/jakob2019.py  11
- colour/recovery/tests/test_jakob2019.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724&bid=18835358)

File: colour/recovery/tests/test_jakob2019.py (hunk 3 above)
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:64]: Codacy found an issue: [Unused variable 'recovered_sd'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

File: colour/recovery/tests/test_jakob2019.py (hunk 4 above)
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:114]: Codacy found an issue: [Unused variable 'derror'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- colour/recovery/tests/test_jakob2019.py  4
- colour/recovery/jakob2019.py  21
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724&bid=18835358)

File: colour/recovery/tests/test_jakob2019.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+                XYZ = sd_to_XYZ(sd, illuminant=D65) / 100
+
+                RGB = XYZ_to_RGB(
+                    XYZ,
+                    D65_xy,
+                    colourspace.whitepoint,
+                    colourspace.XYZ_to_RGB_matrix,
+                )
+
+                recovered_sd, error = RGB_to_sd_Jakob2019(
```
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:64]: Codacy found an issue: [Unused variable 'recovered_sd'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

File: colour/recovery/tests/test_jakob2019.py [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+        # error_function will not align these for us.
+        shape = DEFAULT_SPECTRAL_SHAPE_JAKOB_2019
+        aligned_cmfs = cmfs.copy().align(shape)
+        illuminant = D65.copy().align(shape)
+        illuminant_XYZ = sd_to_XYZ(D65)
+        illuminant_XYZ /= illuminant_XYZ[1]
+
+        for coefficients in coefficient_list:
+            error, derror, R, XYZ, Lab = error_function(
```
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:114]: Codacy found an issue: [Unused variable 'derror'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- colour/recovery/tests/test_jakob2019.py  4
- colour/recovery/jakob2019.py  21
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724&bid=18835358)

File: colour/recovery/tests/test_jakob2019.py [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
```
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:117]: Codacy found an issue: [Unused variable 'derror'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

File: colour/recovery/tests/test_jakob2019.py [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+
+        shutil.rmtree(self._temporary_directory)
+
+    def test_interpolator(self):
+        """
+        Tests the entirety of the
+        :class:`colour.recovery.jakob2019.Jakob2019Interpolator`class.
+        """
+        interpolator = Jakob2019Interpolator()
+        interpolator.generate(sRGB, cmfs, D65, 4, 4, False)
```
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:215]: Codacy found an issue: [Too many positional arguments for method call](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

File: colour/recovery/tests/test_jakob2019.py [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+                XYZ = sd_to_XYZ(sd, illuminant=D65) / 100
+
+                RGB = XYZ_to_RGB(
+                    XYZ,
+                    D65_xy,
+                    ProPhotoRGB.whitepoint,
+                    ProPhotoRGB.XYZ_to_RGB_matrix,
+                )
+
+                recovered_sd, error = RGB_to_sd_Jakob2019(
```
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:67]: Codacy found an issue: [Unused variable 'recovered_sd'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- colour/recovery/tests/test_jakob2019.py  4
- colour/recovery/jakob2019.py  21
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724&bid=18835358)

File: colour/recovery/tests/test_jakob2019.py (hunk 8 above)
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:215]: Codacy found an issue: [Too many positional arguments for method call](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

File: colour/recovery/tests/test_jakob2019.py (hunk 7 above)
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:117]: Codacy found an issue: [Unused variable 'derror'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

File: colour/recovery/tests/test_jakob2019.py (hunk 9 above)
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:67]: Codacy found an issue: [Unused variable 'recovered_sd'](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

File: colour/recovery/jakob2019.py [hunk 10]
```diff
@@ -0,0 +1,729 @@
...
+
+    def generate(self,
+                 colourspace,
+                 cmfs,
+                 illuminant,
+                 resolution,
+                 verbose=True):
+        """
+        Create a lookup table for a given *RGB* colourspace and of a given
+        resolution.
```
Comment (KelSolaar) [inline colour/recovery/jakob2019.py:616]: Codacy found an issue: [Trailing whitespace](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- colour/recovery/jakob2019.py  21
- colour/recovery/tests/test_jakob2019.py  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724&bid=18835358)

File: colour/recovery/tests/test_jakob2019.py [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+
+        shutil.rmtree(self._temporary_directory)
+
+    def test_interpolator(self):
+        """
+        Tests the entirety of the
+        :class:`colour.recovery.jakob2019.Jakob2019Interpolator`class.
+        """
+        interpolator = Jakob2019Interpolator()
+        interpolator.generate(sRGB, cmfs, D65, 4, 4, verbose=False)
```
Comment (KelSolaar) [inline colour/recovery/tests/test_jakob2019.py:215]: Codacy found an issue: [Argument 'verbose' passed by position and keyword in method call](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724)

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- colour/recovery/tests/test_jakob2019.py  4
- colour/recovery/jakob2019.py  21
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724&bid=18835358)

**Comment by KelSolaar** [comment]
There are a few minor stylistic issues that I will take care of, other than that it LGTM at first glance! FYI, I will push on-top of your branch.

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- colour/recovery/jakob2019.py  21
- colour/recovery/tests/test_jakob2019.py  4
         

Clones added
============
- colour/recovery/jakob2019.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724&bid=18835358)

**Comment by KelSolaar** [comment]
@MichaelMauderer : Did you have a chance to look at that? It is quite beefy, I have some small polishing bits I would like to do but nothing that would prevent merging it!

@enneract : Would it be possible to add the RGB to XYZ conversion for the generic `XYZ_to_sd` definition, please?

File: colour/recovery/jakob2019.py [hunk 12]
```diff
@@ -680,49 +751,79 @@ def optimize(L, coefficients_0):
...
+                raise ValueError(
+                    'Bad magic number, this is likely not the right file type!'
+                )
+
+            self.resolution = struct.unpack('i', coeff_file.read(4))[0]
+            self.scale = np.fromfile(
+                coeff_file, count=self.resolution, dtype=np.float32)
+            self.coefficients = np.fromfile(
+                coeff_file,
+                count=3 * self.resolution ** 3 * 3,
```
Comment (MichaelMauderer) [inline colour/recovery/jakob2019.py:808]: Missing space, and would also re commend adding braces to clarify precedence.

File: colour/algebra/common.py [hunk 13]
```diff
@@ -168,3 +173,31 @@ def spow(a, p):
...
+    -------
+    array_like
+        Array :math:`x` after *smoothstep* sigmoid-like function evaluation.
+    """
+
+    x = as_float_array(x)
+
+    i = np.clip((x - a) / (b - a), 0, 1) if clip else x
+
+    return i ** 2 * (3 - 2 * i)
```
[the remark this answers is not in the document]
Comment (KelSolaar) [reply colour/algebra/common.py:203]: Thanks! Addressed in https://github.com/colour-science/colour/tree/feature/jakob2019

File: colour/recovery/jakob2019.py [hunk 14]
```diff
@@ -39,50 +40,100 @@
...
 """
+
 ACCEPTABLE_DELTA_E = 2.4 / 100  # 1% of JND
 """
+Acceptable *perceptual* distance in the *CIE L\\*a\\*b\\** colourspace.
+
+Notes
+-----
+Because *Jakob and Hanika (2019)* uses :math:`\\Delta E_{76}` in the
+*CIE L\\*a\\*b\\** colourspace, the optimization errors are maybe not ideal.
```
[the remark this answers is not in the document]
Comment (KelSolaar) [reply colour/recovery/jakob2019.py:63]: Yes! Wrote a fattier paragraph in https://github.com/colour-science/colour/tree/feature/jakob2019

File: colour/recovery/jakob2019.py (hunk 12 above)
Comment (KelSolaar) [reply colour/recovery/jakob2019.py:808]: Updated too!

File: colour/recovery/jakob2019.py [hunk 15]
```diff
@@ -39,50 +40,100 @@
...
     return SpectralDistribution(R, wl, name=name)
 
 
 class StopMinimizationEarly(Exception):
     """
     The exception used to stop :func:`scipy.optimize.minimize` once the
-    value of the minimized function is small enough. Currently *SciPy* doesn't
-    provide a better way of doing it.
+    value of the minimized function is small enough. *SciPy* doesn't currently
+    offer a better way of doing it.
```
[the remark this answers is not in the document]
Comment (KelSolaar) [reply colour/recovery/jakob2019.py:129]: It should be transparent as the raised exception is caught internally and never surfaced to the user, it is mostly an implementation detail. Correct me if wrong @enneract!

**Comment by KelSolaar** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- colour/recovery/tests/test_jakob2019.py  4
- colour/recovery/jakob2019.py  21
         

Clones added
============
- colour/recovery/jakob2019.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/colour-science/colour/pullRequest?prid=5753724&bid=18835358)

**Comment by KelSolaar** [comment]
Thanks @enneract!

Merging! :) I will do some polishing on the side.

**Pull request merged**